## HTTP-Request methods

Source: https://www.medienpalast.net/glossar/begriff/http-request/

HTTP (Hypertext Transfer Protocol Request) defines a set of request methods to indicate the purpose of the request and what is expected if the request is successful.

The requests come from Client, typically a web browser and would be sent to a Server, to request resources. These resources can be websites, images, videos or have another format, that is available over the internet. The communication between Client and Server takes place over the HTTP-Protocol.


## HTTP-Methods

The most common HTTP methods are:

- GET: to retrieve information from the server

- POST: to serve to transmit data to the server

- PUT: to update the data

- DELETE: the delete resources


### Components

The first fine contains:

- the URL of the requested resource
- the HTTP method
- the HTTP version

Following the first line are the HEADER, which provide additional information about the request, such as the type of client or the accepted formats of the returned data.

POST and PUT requests also include a BODY containing the data to be transmitted. This body can be in various formats such as JSON, XML, or form data. The body is particularly important because it contains the information the server needs to process to fulfill the request.

### Why are HTTP Requests important?

They aim to provide the desired information quickly and efficiently. If an HTTP request fails due to an error or an invalid request, this often has an immediate impact on users: an error message is displayed, and the user does not receive the requested page.

### Error Handling

Servers can return various status codes to indicate the state of a request. For example, a successful GET request returns a 200 status code, while a 404 error indicates that the requested resource could not be found. These status messages are important for developers to identify and resolve errors.

In [3]:
import requests

In [4]:
req = requests.get("https://api.github.com")

In [5]:
# check if retrieve information is successful

req.status_code

200

In [6]:
# the retrieved information in format byte

req.content

b'{"current_user_url":"https://api.github.com/user","current_user_authorizations_html_url":"https://github.com/settings/connections/applications{/client_id}","authorizations_url":"https://api.github.com/authorizations","code_search_url":"https://api.github.com/search/code?q={query}{&page,per_page,sort,order}","commit_search_url":"https://api.github.com/search/commits?q={query}{&page,per_page,sort,order}","emails_url":"https://api.github.com/user/emails","emojis_url":"https://api.github.com/emojis","events_url":"https://api.github.com/events","feeds_url":"https://api.github.com/feeds","followers_url":"https://api.github.com/user/followers","following_url":"https://api.github.com/user/following{/target}","gists_url":"https://api.github.com/gists{/gist_id}","hub_url":"https://api.github.com/hub","issue_search_url":"https://api.github.com/search/issues?q={query}{&page,per_page,sort,order}","issues_url":"https://api.github.com/issues","keys_url":"https://api.github.com/user/keys","label_sea

In [7]:
req.text

'{"current_user_url":"https://api.github.com/user","current_user_authorizations_html_url":"https://github.com/settings/connections/applications{/client_id}","authorizations_url":"https://api.github.com/authorizations","code_search_url":"https://api.github.com/search/code?q={query}{&page,per_page,sort,order}","commit_search_url":"https://api.github.com/search/commits?q={query}{&page,per_page,sort,order}","emails_url":"https://api.github.com/user/emails","emojis_url":"https://api.github.com/emojis","events_url":"https://api.github.com/events","feeds_url":"https://api.github.com/feeds","followers_url":"https://api.github.com/user/followers","following_url":"https://api.github.com/user/following{/target}","gists_url":"https://api.github.com/gists{/gist_id}","hub_url":"https://api.github.com/hub","issue_search_url":"https://api.github.com/search/issues?q={query}{&page,per_page,sort,order}","issues_url":"https://api.github.com/issues","keys_url":"https://api.github.com/user/keys","label_sear

In [9]:
# information in JSON-Format

req.json()

{'current_user_url': 'https://api.github.com/user',
 'current_user_authorizations_html_url': 'https://github.com/settings/connections/applications{/client_id}',
 'authorizations_url': 'https://api.github.com/authorizations',
 'code_search_url': 'https://api.github.com/search/code?q={query}{&page,per_page,sort,order}',
 'commit_search_url': 'https://api.github.com/search/commits?q={query}{&page,per_page,sort,order}',
 'emails_url': 'https://api.github.com/user/emails',
 'emojis_url': 'https://api.github.com/emojis',
 'events_url': 'https://api.github.com/events',
 'feeds_url': 'https://api.github.com/feeds',
 'followers_url': 'https://api.github.com/user/followers',
 'following_url': 'https://api.github.com/user/following{/target}',
 'gists_url': 'https://api.github.com/gists{/gist_id}',
 'hub_url': 'https://api.github.com/hub',
 'issue_search_url': 'https://api.github.com/search/issues?q={query}{&page,per_page,sort,order}',
 'issues_url': 'https://api.github.com/issues',
 'keys_url': '

In [10]:
req.headers

{'Date': 'Sun, 27 Sep 2026 11:48:11 GMT', 'Cache-Control': 'public, max-age=60, s-maxage=60', 'Vary': 'Accept,Accept-Encoding, Accept, X-Requested-With', 'ETag': '"4f825cc84e1c733059d46e76e6df9db557ae5254f9625dfe8e1b09499c449438"', 'x-github-api-version-selected': '2022-11-28', 'Access-Control-Expose-Headers': 'ETag, Link, Location, Retry-After, X-GitHub-OTP, X-RateLimit-Limit, X-RateLimit-Remaining, X-RateLimit-Used, X-RateLimit-Resource, X-RateLimit-Reset, X-OAuth-Scopes, X-Accepted-OAuth-Scopes, X-Poll-Interval, X-GitHub-Media-Type, X-GitHub-SSO, X-GitHub-Request-Id, Deprecation, Sunset, Warning', 'Access-Control-Allow-Origin': '*', 'Strict-Transport-Security': 'max-age=31536000; includeSubdomains; preload', 'X-Frame-Options': 'deny', 'X-Content-Type-Options': 'nosniff', 'X-XSS-Protection': '0', 'Referrer-Policy': 'origin-when-cross-origin, strict-origin-when-cross-origin', 'Content-Security-Policy': "default-src 'none'", 'Server': 'github.com', 'Content-Type': 'application/json; ch

In [13]:
# try GET-method with parameters

params = {"q": "requests+language:python"}
response = requests.get("http://api.github.com/search/repositories", params=params)

In [16]:
response.json()

{'total_count': 422,
 'incomplete_results': False,
 'items': [{'id': 33210074,
   'node_id': 'MDEwOlJlcG9zaXRvcnkzMzIxMDA3NA==',
   'name': 'secrules-language-evaluation',
   'full_name': 'SpiderLabs/secrules-language-evaluation',
   'private': False,
   'owner': {'login': 'SpiderLabs',
    'id': 508521,
    'node_id': 'MDEyOk9yZ2FuaXphdGlvbjUwODUyMQ==',
    'avatar_url': 'https://avatars.githubusercontent.com/u/508521?v=4',
    'gravatar_id': '',
    'url': 'https://api.github.com/users/SpiderLabs',
    'html_url': 'https://github.com/SpiderLabs',
    'followers_url': 'https://api.github.com/users/SpiderLabs/followers',
    'following_url': 'https://api.github.com/users/SpiderLabs/following{/other_user}',
    'gists_url': 'https://api.github.com/users/SpiderLabs/gists{/gist_id}',
    'starred_url': 'https://api.github.com/users/SpiderLabs/starred{/owner}{/repo}',
    'subscriptions_url': 'https://api.github.com/users/SpiderLabs/subscriptions',
    'organizations_url': 'https://api.git

In [17]:
response.json()['items'][0]

{'id': 33210074,
 'node_id': 'MDEwOlJlcG9zaXRvcnkzMzIxMDA3NA==',
 'name': 'secrules-language-evaluation',
 'full_name': 'SpiderLabs/secrules-language-evaluation',
 'private': False,
 'owner': {'login': 'SpiderLabs',
  'id': 508521,
  'node_id': 'MDEyOk9yZ2FuaXphdGlvbjUwODUyMQ==',
  'avatar_url': 'https://avatars.githubusercontent.com/u/508521?v=4',
  'gravatar_id': '',
  'url': 'https://api.github.com/users/SpiderLabs',
  'html_url': 'https://github.com/SpiderLabs',
  'followers_url': 'https://api.github.com/users/SpiderLabs/followers',
  'following_url': 'https://api.github.com/users/SpiderLabs/following{/other_user}',
  'gists_url': 'https://api.github.com/users/SpiderLabs/gists{/gist_id}',
  'starred_url': 'https://api.github.com/users/SpiderLabs/starred{/owner}{/repo}',
  'subscriptions_url': 'https://api.github.com/users/SpiderLabs/subscriptions',
  'organizations_url': 'https://api.github.com/users/SpiderLabs/orgs',
  'repos_url': 'https://api.github.com/users/SpiderLabs/repos',


In [21]:
# note: with POST-method, the link must be correct so that the data can be posted

req = requests.post('http://httpbin.org/post',data = {'post-method':'testfromme'})
req.json()

# data that is sent to server is stored in form

{'args': {},
 'data': '',
 'files': {},
 'form': {'post-method': 'testfromme'},
 'headers': {'Accept': '*/*',
  'Accept-Encoding': 'gzip, deflate, zstd',
  'Content-Length': '22',
  'Content-Type': 'application/x-www-form-urlencoded',
  'Host': 'httpbin.org',
  'User-Agent': 'python-requests/2.33.1',
  'X-Amzn-Trace-Id': 'Root=1-6ab9070b-3a1b606e0b2352fc579ea941'},
 'json': None,
 'origin': '89.246.102.198',
 'url': 'http://httpbin.org/post'}

In [22]:
my_json = {"post-method":"test-json"}
req = requests.post('http://httpbin.org/post',json = my_json)
req.json()

{'args': {},
 'data': '{"post-method": "test-json"}',
 'files': {},
 'form': {},
 'headers': {'Accept': '*/*',
  'Accept-Encoding': 'gzip, deflate, zstd',
  'Content-Length': '28',
  'Content-Type': 'application/json',
  'Host': 'httpbin.org',
  'User-Agent': 'python-requests/2.33.1',
  'X-Amzn-Trace-Id': 'Root=1-6ab91147-03a3c52358b9f88d5b0a71d8'},
 'json': {'post-method': 'test-json'},
 'origin': '89.246.102.198',
 'url': 'http://httpbin.org/post'}

In [23]:
with open("http_test.txt", "w") as file:
    file.write("http_test")

In [24]:
my_file = {"file": open("http_test.txt", "rb")}
req = requests.post("http://httpbin.org/post",files = my_file)
req.json()

{'args': {},
 'data': '',
 'files': {'file': 'http_test'},
 'form': {},
 'headers': {'Accept': '*/*',
  'Accept-Encoding': 'gzip, deflate, zstd',
  'Content-Length': '158',
  'Content-Type': 'multipart/form-data; boundary=38c7243e9733c8bd2abf4d8438b1bc39',
  'Host': 'httpbin.org',
  'User-Agent': 'python-requests/2.33.1',
  'X-Amzn-Trace-Id': 'Root=1-6ab911d8-7f40810031106b5550ee7c24'},
 'json': None,
 'origin': '89.246.102.198',
 'url': 'http://httpbin.org/post'}

In [26]:
# put-method

req = requests.put("http://httpbin.org/put",data = {"put-method":"test-put-method"})
req.json()

{'args': {},
 'data': '',
 'files': {},
 'form': {'put-method': 'test-put-method'},
 'headers': {'Accept': '*/*',
  'Accept-Encoding': 'gzip, deflate, zstd',
  'Content-Length': '26',
  'Content-Type': 'application/x-www-form-urlencoded',
  'Host': 'httpbin.org',
  'User-Agent': 'python-requests/2.33.1',
  'X-Amzn-Trace-Id': 'Root=1-6ab91234-384ed5d415d413c8263a483e'},
 'json': None,
 'origin': '89.246.102.198',
 'url': 'http://httpbin.org/put'}

In [30]:
# delete-method
url = "http://httpbin.org/get"
req = requests.delete("http://httpbin.org/delete")
req = requests.delete(url)

In [31]:
from getpass import getpass
requests.get('http://api.github.com/user',auth = ('username',getpass()))

<Response [401]>

The HTTP error 401 Unauthorized indicates that a request was sent without valid authentication or that authentication failed. This means that the user did not provide the necessary credentials to access the requested resource.

### HTTP Status-Code

Source: https://restfulapi.net/http-status-codes/

HTTP defines these standard status codes that can be used to convey the results of a client’s request. The status codes are divided into five categories.

- 1xx: Informational – Communicates transfer protocol-level information.
- 2xx: Success – Indicates that the client’s request was accepted successfully.
- 3xx: Redirection – Indicates that the client must take some additional action in order to complete their request.
- 4xx: Client Error – This category of error status codes points the finger at clients.
- 5xx: Server Error – The server takes responsibility for these error status codes.

